[WIP] Initial draft of Sidecar Service and Managed Proxy deprecation docs

banks opened the pull request:
Not sure if I'll break up all the docs into separate PRs or iterate on this one but I'll list TODO for all docs changes on needed on this branch.

## TODO

 - [x] Changes to Service Definition page related to sidecar service
 - [x] Changes to Service Definition page related to deprecated managed proxy
 - [x] Changes to Agent config for sidecar service (auto port range)
 - [x] Changes to Agent config for deprecated all managed proxy settings
 - [x] Page on Envoy support including custom listener config and envoy command.
 - [x] Document `connect envoy` sub command
 - [x] Document `connect proxy -sidecar-for` option
 - [x] Grep all references to managed proxies or proxy tokens
 - [x] Document gRPC config/CLI options
 - [x] Changes to getting started Connect guide to use sidecar service and link out to K8s/docker
 - [ ] K8s getting started guide? @mitchellh are you expecting to do this? Can be separate
 - [ ] Maybe, get docker-compose getting started guide up (probably separate PR if/when that is ready
 - [ ] Other refactors after a read through?
 - [x] API docs for new /agent/service/foo

banks commented:
Should document sidecar service registration lifecycle

mitchellh commented on website/source/docs/connect/proxies/envoy.md:
```diff
@@ -0,0 +1,192 @@
+  consul-envoy -sidecar-for web
+3f213a3cf9b7583a194dd0507a31e0188a03fc1b6e165b7f9336b0b1bb2baccb
+$ docker run --rm -d --network host \
+  consul-envoy -sidecar-for db -admin-bind localhost:19001
```
I would just make a note somewhere in the docs why this `-admin-bind` flag is necessary.

banks commented:
Envoy guide is complete although probably needs review/edit.

I also deployed a complete version of the docs to a netlify site to allow easier review.

For example you can see the Envoy docs here: https://consul-docs-banks-dev.netlify.com/docs/connect/proxies/envoy.html

pearkes commented on website/source/docs/commands/connect/envoy.html.md.erb:
```diff
@@ -0,0 +1,178 @@
+
+The default behaviour is to generate the necessary bootstrap configuration for
+envoy based on the environment variables and options provided and by taking to
+the local Consul agent. It calls `exec` the envoy binary with that configuration
```
Grammar here maybe? "`exec` [on] the envoy binary"?.

Also thinking there should the note about PATH here up-front. I think the subcommand should have an obvious sentence above the fold in it that notes that it doesn't embed Envoy but is just exec'ing.

banks replied:
I changed this to 

> It `exec`s an external Envoy binary with that
configuration leaving the Envoy process running in the foreground.

I left out `$PATH` just because that's only the default and a custom binary can be used which I don't want to get into in this intro paragraph but added `external` to call that out more clearly.

banks commented:
Note: for some reason I was under the impression that Envoy prefers to lower case the name in general as it appears that way so often and made a conscious effort to do that in docs... Turns out they don't and even on their own website they capitalise it normally so I've been through all the changes here and re capitalized `Envoy` everywhere it's not being used in a command/config/url etc.

banks commented:
Updated the Connect getting started guide: https://consul-docs-banks-dev.netlify.com/intro/getting-started/connect.html

And created a new `Getting Started With Connect` section (https://consul-docs-banks-dev.netlify.com/docs/connect/index.html#getting-started-with-connect) in the Connect index that lists all the various ways to try things out. That might need an updated link when the Connect + Kube stuff lands.

Input welcome - I don't think the extra few steps of running proxies really made that guide any worse. I had to reword things a few times to not be confusing and I ran through and actually tested all the commands verbatim work locally etc. Input welcome though @mitchellh @pearkes 

I *think* that's everything that's essential but would really appreciate anyone's input on ways we could improve overall structure/presentation (or things i've missed). They don't necessarily need to block 1.3 though.

kaitlincarter-hc commented:
I was working on some ideas to make this paragraph flow a bit smoother:

Original - "The simplest way to try out Envoy with Consul locally is using Docker. Envoy doesn't release binaries outside of their official Docker image. If you can build Envoy directly then the consul connect envoy command command can be used directly on your local machine to start Envoy, however for this guide we'll use the Docker image."

Updated - "The quickest way to try out Envoy with Consul is by using Docker locally. Note, Envoy only releases binaries as official Docker image. We will be using the official Docker image in this guide. Alternatively, you can build Envoy directly then use the consul connect envoy command on your local machine to start Envoy."

kaitlincarter-hc commented:
Some thoughts on the following paragraph:

"To get a simple example working, run a single agent in -dev mode using the latest Docker image. It's possible to run the agent on the Docker host but it complicates the rest of the steps below due to networking config on different platforms. This guide will start all containers using Docker host network which is not a good simulation of a production setup, but makes the following steps much simpler."

Revised opening sentence: "To get a simple example working we will run a single agent in development mode."

Comments:
“It's possible to run the agent on the Docker host but it complicates the rest of the steps below due to networking config on different platforms.” - This sentence confuses me. It might be nice to clearly state at the beginning how many docker containers we will be using. 

"This guide will start all containers using Docker host network which is not a good simulation of a production setup, but makes the following steps much simpler.” This feels a bit like a side note. What if we have a section at the top that details what the guide environment will look like so we can remove it from the guide.

kaitlincarter-hc commented:
Simple is used often in the guide, here are some recommended alternatives: straightforward, primitive, uncomplicated, plain.

kaitlincarter-hc commented:
Can we move supported versions to the top? I feel like its a prerequisite.

banks commented:
@kaitlincarter-hc thanks for the notes I think those are great suggestions and I'll rework thinks a little. Some comments inline.

For future reference, PR comments can be attached to specific lines of the diff if you click on the `Files changed` tab at the top and that way the conversation about each suggestion can be threaded. This is fine for now just FYI.

> Structure notes: As mentioned in an earlier comment, it might be nice to have a environment description at the beginning of the guide

Makes sense yeah describing the end goal separately from the steps needed is a good plan.

> Could we move the Advanced Configuration to a new guide?

Hmm yeah I was kinda cheating a bit throwing this all in one place. Our normal pattern is to keep docs pages like this more on the "reference documentation" side (which the advanced config section is) and have separate "Guides" for walk-throughs. I think out one big pot of "guides" though is getting unmanageable - there are lots of things in there that range from the "ACL Guide" which is not at all a guide and is our only ACL reference doc! to "general advice about X" documents. It could probably do with more organising and I know there are plans to move some of that elsewhere so I'm avoiding touching that for now!

I'd kinda like to have a more specific place for "Getting started with connect" to be honest. I might try adding a new section in the Connect docs rather than just the list I added here to the Connect index page.

Let me go and work on the envoy part again.

banks commented:
@kaitlincarter-hc I've moved out the Envoy getting started section into a separate guide based on the structural pointers you gave. I also added some more reference info that is important and I missed on the full Envoy page.

Feedback welcome, I think I'm OK with this overall structure for now for release - we can iterate on the guides and new user experience from here.

banks commented:
Rebased on master now `f-envoy` has landed there. Have a couple more TODOs and feed back incoming:

 - [ ] Document sidecar service lifecycle
 - [ ] Envoy command menu capitalised

kaitlincarter-hc commented:
@banks Thanks for the tip on attaching comments to specific lines, I am new to this workflow! 

I really like the separate guide and reference document! The getting started guide is really clear and easy to follow now. I'd say it's ready to go live and we can revisit updates later. 

I know the guide section in the docs is a bit messy, so I understand why you would want to avoid adding more. However, don't worry about it, I am working on untangling the docs/guides/lms.

banks commented on website/source/docs/agent/services.html.md:
```diff
@@ -93,39 +94,92 @@ meta object in node definition.
+[anti-entropy syncs](/docs/internals/anti-entropy.html) and deregistration.
+
+A service can have an associated health check. This is a powerful feature as
+it allows a web balancer to gracefully remove failing nodes, a database
```
Heh no that's odd phrasing but good catch since htis is merge fail - this is copy-paste from the ### Checks section below which didn't change. I'll fix it up.

banks commented on website/source/docs/agent/services.html.md:
```diff
@@ -93,39 +94,92 @@ meta object in node definition.
+node has any failing system-level check, the DNS interface will omit that
+node from any service query.
+
+The check must be of the script, HTTP, TCP or TTL type. If it is a script type,
```
Yeah they were missed out when added, note that this para is also merge-fail-dupe of the old Checks section which is still below. I'll fix up there.
